### PR TITLE
MODOAIPMH-300: Not all instances ids are saved for initial harvest

### PR DESCRIPTION
**Pull request by IlliaDaliek**
Jira: https://issues.folio.org/browse/MODOAIPMH-300
Jira: https://issues.folio.org/browse/MODOAIPMH-301

### PURPOSE
Fix issue with the hanging of srs response handler which caused the harvest to be canceled. 
The root cause is that when srs responds with a large response body it may hang for an unknown reason and the body is not written to the response, therefore we can get the body and invoke body handler which causes the code hanging and canceling harvesting as result.

### APPROCH

Set the idle timeout of srs request to 20 seconds (instead of 5000) in order to invoke the exception handler (by causing timeout exception) when there a problem with receiving response body from SRS. Make a retry of srs call in the exception handler.

### CODE REVIEW

File: src/test/java/org/folio/rest/impl/OaiPmhImplTest.java
Comment (IlliaDaliek) [inline src/test/java/org/folio/rest/impl/OaiPmhImplTest.java:2202]: Unit test for re-requesting srs was written before this PR
It is called shouldRerequestSrsWhenItRespondedWith500_whenGetOaiRecordsMarc21WithHoldingsWithErrorResponseAndThenSuccessResponseFromSrs

File: src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java
Comment (craigmcnally) [inline src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java:31]: Since units were a problem here, it may be a good idea to incorporate the units into the variable names... e.g. `idleTimeoutSec`, `connTimeoutMs`

File: src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java
Comment (craigmcnally) [inline src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java:70]:
```suggestion
```

File: src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java
Comment (craigmcnally) [inline src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java:442]:
```suggestion
```

File: src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java
Comment (craigmcnally) [inline src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java:44]: Instead of hardcoding these timeout values, can we make them constants?

File: src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java
Comment (craigmcnally) [inline src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java:74]: what is the significance of error 1234?

File: src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java
Comment (craigmcnally) [inline src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java:77]: See comment above about using constants instead of integer literals

File: src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java
Comment (craigmcnally) [inline src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java:725]: I think it's misleading to always report 400 status here...   It's probably better to not report a status than to potentially report the wrong one.

Comment (craigmcnally) [inline src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java:725]: If I read this right, the exception handler provided here will only be invoked upon request exceptions (`request.exceptionHandler(...)`).  In this case there is no status code.  Furthermore, the message provided `Error in SRS response` is inaccurate.  `Error making SRS request` would be more accurate.

File: src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java
Comment (craigmcnally) [inline src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java:733]: It's probably better to log this before calling this method.  The way this currently stands, in some cases you might log a message at ERROR level, then a different (also inaccurate - see earlier comments) message at WARN level.

File: src/main/java/org/folio/oaipmh/Constants.java
Comment (craigmcnally) [inline src/main/java/org/folio/oaipmh/Constants.java:26]: `httpRequestRetryAttempts` is a little misleading - The module makes several HTTP requests.  This configuration only applies to the call to SRS.

File: src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java
Comment (craigmcnally) [reply src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java:74]:
```suggestion
      logger.error("Exception making SRS request: " + e.getMessage(), e);
```

File: pom.xml
Comment (craigmcnally) [inline pom.xml:79]: I didn't see this mentioned in the PR description... is this a necessary part of this PR?  If so, can you please explain why?

Comment (IlliaDaliek) [reply pom.xml:79]: When I had a discussion with Adam and Julian they suggested to use the latest vertx version with major version 3 (3.9.6).
But dependencies with the FOLIO prefix for 3.9.6 don't exist. So we changed the vertx version property value to be 3.9.6 and left some dependencies with 3.9.4-FOLIO.

File: src/main/java/org/folio/oaipmh/Constants.java
Comment (IlliaDaliek) [reply src/main/java/org/folio/oaipmh/Constants.java]: Okay, it makes sense. I added the "srs" prefix to the property name.

File: src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java
Comment (IlliaDaliek) [reply src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java]: Agree, corrected.

File: src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java
Comment (IlliaDaliek) [reply src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java]: Done

File: src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java
Comment (IlliaDaliek) [reply src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java]: Thanks for the note, initially this log was added just for testing of implemented logic work. I deleted it since we have loggin in the "srs retry method".

File: src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java
Comment (IlliaDaliek) [reply src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java]: Yes, done.

File: src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java
Comment (IlliaDaliek) [reply src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java]: Agree, corrected.

**Review by MikhailFokanov — APPROVED** [review]
Looks good.

**Review by illia-borysenko-ebsco — APPROVED** [review]

File: pom.xml
Comment (craigmcnally) [reply pom.xml:79]: In the future, please add this context to the PR description so that it's clear why the change is being made.  Without this information the PR reviewer is left wondering if the change was intentional, or a mistake.

File: src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java
Comment (craigmcnally) [inline src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java:26]: why no units on this one?

File: src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java
Comment (craigmcnally) [inline src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java:75]: I realize this was already there and you're just using it but `GLOBAL_PATH` is a lousy name for this.  Something like `SOURCE_RECORDS_PATH` is IMO much more appropriate

**Review by craigmcnally — DISMISSED** [review]
A couple minor comments related to variable names.  Approving now, but please take my comments into consideration.

File: src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java
Comment (craigmcnally) [reply src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java]: Can we remove this extra newline?

File: src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java
Comment (craigmcnally) [reply src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java]: I still see you passing a hard-coded 400 into this method in the case where there is no response code - because the error occurred while writing the request.  Changing the log level from WARN to DEBUG is not sufficient to fix this IMO.

**Review by craigmcnally — CHANGES_REQUESTED** [review]
Please address the issues raised about the logging in retry method

**Comment by craigmcnally** [comment]
@IlliaDaliek I noticed that you've introduce new changes here after the PR was approved.  Please re-request reviews from those that previously approved.  This feels a little like scope creep here.  Strictly speaking this probably should have happened in a separate PR, probably with its own JIRA.  Given that the changes are already here, we might as well go with it.

**Comment by IlliaDaliek** [comment]
> @IlliaDaliek I noticed that you've introduce new changes here after the PR was approved. Please re-request reviews from those that previously approved. This feels a little like scope creep here. Strictly speaking this probably should have happened in a separate PR, probably with its own JIRA. Given that the changes are already here, we might as well go with it.

@craigmcnally, there no serious changes, I moved srs idle timeout to module settings as Mikhail asked me to do that. As well when I verified the "schema update testing" ticket I noticed incorrect logs and fixed them as well in the current PR. And among the code the laterals were used instead of already defined constants, I corrected this too.

File: src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java
Comment (IlliaDaliek) [reply src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java]: > I still see you passing a hard-coded 400 into this method in the case where there is no response code - because the error occurred while writing the request. Changing the log level from WARN to DEBUG is not sufficient to fix this IMO.
@craigmcnally, it's not correct. 
There re-request can be triggered by two reasons: 
1) when the srs returns a full response but with incorrect status code
2) when there a problem in srs with passing the body and the timeout exception is thrown (or any else potential http exceptions).
These two cases are different situations and they should have different log messages. I corrected it and we didn't pass the hardcoded "400" value. 
image

File: src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java
Comment (IlliaDaliek) [reply src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java:32]: I added a test.

**Review by MikhailFokanov — APPROVED** [review]

**Comment by craigmcnally** [comment]
> > @IlliaDaliek I noticed that you've introduce new changes here after the PR was approved. Please re-request reviews from those that previously approved. This feels a little like scope creep here. Strictly speaking this probably should have happened in a separate PR, probably with its own JIRA. Given that the changes are already here, we might as well go with it.
> 
> @craigmcnally, there no serious changes, I moved srs idle timeout to module settings as Mikhail asked me to do that. As well when I verified the "schema update testing" ticket I noticed incorrect logs and fixed them as well in the current PR. And among the code the laterals were used instead of already defined constants, I corrected this too.

OK thanks.  I can't stress this enough how important it is to communicate these details in the PR.  Otherwise, it's very hard to understand what's going on, and why.

**Review by craigmcnally — APPROVED** [review]

**Pull request merged**
